### PR TITLE
Add long tap to right click support

### LINKED ISSUE #5200: Add ability to map long taps to right click
Right now there is no way to trigger right clicks on a touch-only device. This needs to be added, and probably set as a default behaviour when using the touch-to-mouse mapping logic.

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes #5200 

Went with a hold delay of 500ms from real device testing, feels pretty well. The implementation behind this was to trigger a right click and "cancel" the held left button (in which it would be released but not trigger a click or otherwise).

In addition, I had to move the button input application to happen only if `e.IsActive != null` rather than always, since the latter would cause left clicks when moving a touch slightly after triggering a right click.

### CODE REVIEW

**Comment by peppy** [comment]
To make this feel good, we're going to need some kind of visual feedback during the operation. I'm not yet sure where this should be done, but may be good to do some initial experimentation and see what you come up with.

Take a look at iOS for the best implementation I've seen to date (actually scaling the target element during the operation). I'm not sure how possible this is to do for us since we haven't really thought about it from the outset. Alternatives would be to show some kind of overlay effect on or under the cursor (windows [does this](https://www.zbrushcentral.com/t/windows-wacom-and-zbrush-wackiness/286378)) which scales or radial fills to show how far through the hold-to-right-click cycle the user is.

I can see this being something game/applications may want to override (ie. osu! side making it part of the menu cursor seems like a good place), so a good start could be exposing this from `InputManager` as a progress bindable, or at very least `Started` and `Ended` events along with `touch_right_click_delay` being `public`.

**Comment by bdach** [comment]
Looking at the implementation from the technical side, I'm not sure I agree with the way these "long press" events are synthesised from normal touches. Both [iOS](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/handling_uikit_gestures/handling_long-press_gestures) and [Android](https://developer.android.com/reference/android/view/View.OnLongClickListener) have dedicated OS pathways for this, and I'm not sure we should be doing our own things rather than using what the OSes provide.

**Comment by frenzibyte** [comment]
> Looking at the implementation from the technical side, I'm not sure I agree with the way these "long press" events are synthesised from normal touches. Both [iOS](https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/handling_uikit_gestures/handling_long-press_gestures) and [Android](https://developer.android.com/reference/android/view/View.OnLongClickListener) have dedicated OS pathways for this, and I'm not sure we should be doing our own things rather than using what the OSes provide.

The one concern I have is we use this gesture whenever touch input is not handled by the held drawable, rather than always using it.

If we were to implement this from OS events it would mean defining the concept of a “gesture” to the o!f input flow until it reaches `InputManager` where we can add visual feedback to it.

I believe it’s better implementing it this way until we require long touch gestures in game components that handle multi-touch input.

**Comment by bdach** [comment]
> If we were to implement this from OS events it would mean defining the concept of a “gesture” to the o!f input flow until it reaches `InputManager` where we can add visual feedback to it.

This is a hypercorrect stance. From a pragmatic standpoint I see no reason to take it this far. You could still proxy the OS "long press gestures" to right clicks for the time being, but the OS would be responsible for actually detecting them.

The reason why I'm not really all that okay with the current proposal is that each OS has its own paradigms and if the OS provides an input flow for long press, we should use that to be consistent with _literally everything else on the platform_ if it's feasible.

**Comment by frenzibyte** [comment]
> > If we were to implement this from OS events it would mean defining the concept of a “gesture” to the o!f input flow until it reaches `InputManager` where we can add visual feedback to it.
> 
> This is a hypercorrect stance. From a pragmatic standpoint I see no reason to take it this far. You could still proxy the OS "long press gestures" to right clicks for the time being, but the OS would be responsible for actually detecting them.

Let me clarify, once the long press gesture is triggered, the held touch will have to be ignored, and not cause a left mouse click (otherwise you could enter beatmaps when wanting to right-click on them in the carousel). This is not feasible without introducing a “cancelled” state to all mouse button input which will be initiated from touch input handlers, and that may require a good amount of effort, as opposed to the current logic in this PR.

I’m willing to give it a try on iOS and see where it leads.

**Comment by bdach** [comment]
Yeah I don't know. I'm uneasy with rolling our own code for press vs long press detection if the OSes already have got it handled somehow. "Cancelled" mouse inputs, or the `IgnoreClick` you have in this diff, don't exactly inspire confidence in me. But maybe that's just me.

**Comment by frenzibyte** [comment]
Also realized this will break any component which requires the user to hold onto (mainly osu! ruleset, holding slider)... Will need to rethink this structure a bit, in addition to trying out OS gestures.

**Comment by peppy** [comment]
Should only be handled on a single touch, if another finger touches it should permanently cancel until all are released. Non-transferrable between fingers. aka, keep it simple.

> * Should right click be activated when holding with one finger and swiping with it before it's triggered? Or should it reset the counter every touch move? Or should it never activate.

Moving the touch beyond a certain amount should cancel it. Probably the same as the drag lenience.

**Comment by frenzibyte** [comment]
Have addressed all of the points above with a bit of a refactor, this feels quite good on thorough testing for me.

File: osu.Framework/Input/InputManager.cs [hunk 1]
```diff
@@ -737,21 +770,61 @@ protected virtual bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
                 }.Apply(CurrentState, this);
             }
 
-            switch (e.IsActive)
+            if (e.IsActive != null)
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Framework/Input/InputManager.cs:773]: It's null when `HandleMouseTouchStateChange` invokes due to a change in touch position only.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Input/InputManager.cs:773]: Right, I misread the conditionals to mean something different to what they actually are doing.

**Comment by peppy** [comment]
Regarding the conversation of using the operating system level gesture / long touch flow: I do agree it is something we should eventually aim for, but it comes with some hard issues that may take a paradigm shift to fix (making our input flow match closer to mobile expectations).

The way it's done in this PR seems like a good initial step which will at least give us the behaviour we want in a generically accepted form. Can iterate at a later point based on user feedback.

**Comment by peppy** [comment]
I *finally* got around to testing this out on some devices. Sorry for the delay.

I've made some visual tweaks as I couldn't see the circle on smaller devices (like the iPhone SE 4.5" or whatever it was).

Remaining issues as I see:

- After starting a drag (ie. cancelling the hold), a new hold should not start until the finger has been lifted. Right now it starts again whenever the input is still
- The leniency seems a bit too low, it's easy to accidentally move the input too much while trying to hold


Both of these issues can be seen in this capture, where I do not lift my finger once:


https://user-images.githubusercontent.com/191335/194473455-507ea9ec-a78e-4384-b3f2-0dae7ac13485.mov

**Review by peppy — CHANGES_REQUESTED** [review]
As proposed

**Comment by frenzibyte** [comment]
> After starting a drag (ie. cancelling the hold), a new hold should not start until the finger has been lifted. Right now it starts again whenever the input is still

This wouldn't be the case when the drag is handled by a drawable. That is the simplest way to go about it, but given that leniency is required as well. Will have it work by checking against touch down position instead.

**Comment by frenzibyte** [comment]
The mentioned issues should be resolved now, haven't managed to test these changes on a physical device due to latest Xcode version being a nuisance (need to update xamarin packages, but internet speed sucks). @peppy would appreciate checking the latest changes and see if it feels better now.

**Comment by peppy** [comment]
Now the position is locked, but it can still be triggered multiple times... and if you move a touch back to the original location it starts triggering again, even though you've already moved away (should not happen).

**Review by peppy — CHANGES_REQUESTED** [review]
As commented

**Comment by NullifiedJosh** [comment]
This could be a different event too. Maybe like the ios force touch has a different design and better user interaction feedback.'

**Comment by frenzibyte** [comment]
> Now the position is locked, but it can still be triggered multiple times... and if you move a touch back to the original location it starts triggering again, even though you've already moved away (should not happen).

Spent time to rewrite the entire logic to work with touch dragging properly, as the previous one wasn't quite cut for it. Should work correctly now.

**Comment by peppy** [comment]
I'll give this some more testing on a device, but the new test coverage looks ample to cover the cases I pointed out 👍 .

File: osu.Framework/Input/InputManager.cs [hunk 2]
```diff
@@ -807,34 +805,151 @@ private void updateTouchMouseLeft(TouchStateChangeEvent e)
...
+
+            touchLongPressPosition = null;
+
+            touchLongPressDelegate.Cancel();
+            touchLongPressDelegate = null;
+
+            TouchLongPressCancelled?.Invoke();
+        }
+
+        // try #2
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Framework/Input/InputManager.cs:865]: Whoops, unintended code. Was basically retrying how to write the logic as it was quite hard to get right, all until I came up with the final uncommented one which is working. Will remove, shouldn't have been pushed.

File: osu.Framework/Input/MouseButtonEventManager.cs [hunk 3]
```diff
@@ -132,7 +139,7 @@ protected override void HandleButtonUp(InputState state, List<Drawable> targets)
             if (targets != null)
                 PropagateButtonEvent(targets, new MouseUpEvent(state, Button, MouseDownPosition));
 
-            if (EnableClick && DraggedDrawable?.DragBlocksClick != true)
+            if (EnableClick && DraggedDrawable?.DragBlocksClick != true && !IgnoreClick)
```
Comment (peppy) [inline osu.Framework/Input/MouseButtonEventManager.cs:142]: Do you happen to remember why this `IgnoreClick` is required when we already have `BlockNextClick`?

Comment (frenzibyte) [reply osu.Framework/Input/MouseButtonEventManager.cs:142]: Didn't notice that one, should be fine to use it.

**Comment by dengr1065** [comment]
Is it planned to make long-press duration follow the accessibility system setting on Android? Looks like you can access it without receiving long-press gesture events: [ViewConfiguration#getLongPressTimeout](https://developer.android.com/reference/android/view/ViewConfiguration#getLongPressTimeout())

**Pull request merged**